### PR TITLE
[FINISHES #128759285] Desenvolver cenários de criação de grupos de pesquisadores através da seleção de pesquisadores e/ou outros grupos

### PR DESCRIPTION
**Pull request by LcsSantana**
Professor,

Resolvi o problema dos vários arquivos.
E, apesar de uma certa dificuldade, consegui internacionalizar os testes.

---
This change is [Reviewable](https://reviewable.io/reviews/pauloborba/rpa/15)

### CODE REVIEW

**Comment by pauloborba** [comment]
Reviewed 5 of 15 files at r1, 28 of 38 files at r2.
Review status: 33 of 48 files reviewed at latest revision, 12 unresolved discussions.

---

*[grails-app/controllers/rpa/ResearchGroupController.groovy, line 31 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYAD4mRMmEfSY_oy0mL:-KYAD4mRMmEfSY_oy0mM:b-xvjdn6) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/grails-app/controllers/rpa/ResearchGroupController.groovy#L31)):*
> ```Groovy
>             notFound()
>             return
>         }//Da true para qualquer string que contenha ao menos um char não alfanúmérico nem espaço.
> ```

deveria ter colocado como constraint no modelo. o resto parece ter sido todo gerado

---

*[test/cucumber/steps/GroupsSteps.groovy, line 29 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYAAsR2wcV4QUkkFMFy:-KYAAsR33lxksjfSRmNg:b-9u43jx) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/steps/GroupsSteps.groovy#L29)):*
> ```Groovy
>     controladorGrupo.save(grupo)
>     controladorGrupo.response.reset()
>     assert ResearchGroup.findByName(arg1).researchers.contains(pesquisadores[0])
> ```

demeter

---

*[test/cucumber/steps/GroupsSteps.groovy, line 39 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYABFQKd44nGcvFNUpA:-KYABFQKd44nGcvFNUpB:bp09h8w) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/steps/GroupsSteps.groovy#L39)):*
> ```Groovy
>     def grupo = new ResearchGroup([name:arg1, researchers: pesq])
>     for(int i = 0; i < ResearchGroup.findByName(arg4).researchers.size(); i++) {
>         grupo.researchers.add(ResearchGroup.findByName(arg4).researchers[i])
> ```

merece extracao de variavel

era para testar o save passando o nome do grupo

---

*[test/cucumber/steps/GroupsSteps.groovy, line 49 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYAC6fBX9MmdnU9KUm4:-KYAC6fBX9MmdnU9KUm5:b-h9hv7) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/steps/GroupsSteps.groovy#L49)):*
> ```Groovy
>     assert ResearchGroup.findByName(arg1).researchers.contains(Researcher.findByName(arg3))
>     assert ResearchGroup.findByName(arg1).researchers.contains(Researcher.findByName(arg4))
>     assert ResearchGroup.findByName(arg1).researchers.contains(Researcher.findByName(arg5))
> ```

deveria verificar se não colocou mais ninguém no grupo

---

*[test/cucumber/steps/GroupsSteps.groovy, line 54 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYACEdY7eEE9BTuAOE7:-KYACEdY7eEE9BTuAOE8:bud1egh) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/steps/GroupsSteps.groovy#L54)):*
> ```Groovy
> //Controller test2
> When(~/^O sistema recebe uma submissao para adicionar o grupo "([^"]*)"$/) { String arg1 ->
>     def grupo = new ResearchGroup([name:arg1, researchers: null])
> ```

deveria testar com os parâmetros da requisição do cliente

---

*[test/cucumber/steps/GroupsSteps.groovy, line 126 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYACdE4BsSa1y75Qgbs:-KYACdE4BsSa1y75Qgbt:b6fd9xf) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/steps/GroupsSteps.groovy#L126)):*
> ```Groovy
>     to GroupListPage
>     at GroupListPage
>     assert ResearchGroup.findByName(arg1).researchers.contains(Researcher.findByName(arg2))
> ```

deveria verificar na lista da view!

---

*[test/cucumber/support/groups.feature, line 1 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYA8gEbTNUupWHjCqS7:-KYA8gEdaOV7hHDbhhMz:b1mfpv6) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/support/groups.feature#L1)):*
> ```Cucumber
> Feature: Desenvolver cenarios de criacao de grupos de pesquisadores atraves da selecao de pesquisadores e ou outros grupos
> ```

não seguindo formato recomendado

---

*[test/cucumber/support/groups.feature, line 5 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYA8xX0l2fpWC0IkvDi:-KYA8xX0l2fpWC0IkvDj:ba31go2) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/support/groups.feature#L5)):*
> ```Cucumber
> Scenario: Criar grupo de pesquisa novo
>   Given O sistema nao contem o grupo "CIn" cadastrado no seu database
>   And O grupo "InteligenciaArtificialCIn" contem "TeresaLudermir" e "PauloAdeodato"
> ```

contem -> contém apenas, para amarrar bem o comportamento mais adiante

coloca espaço entre os nomes dos professores

---

*[test/cucumber/support/groups.feature, line 21 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYA9jBYixRW1SSUDekk:-KYA9jBYixRW1SSUDekl:b-cbabys) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/support/groups.feature#L21)):*
> ```Cucumber
> Scenario: Criar grupo de pesquisa a partir de pesquisadores em outros grupos
>   Given Eu estou na pagina de Criacao de Grupos
>   And O pesquisador "VeronicaTeichrieb" com o cpf "51111111111" pertence ao grupo "RealidadeAumentada"
> ```

não entendi bem porque essa necessidade de registrar que pertence a um grupo... já que pode pertencer a vários, e isso não deveria impactar o comportamento da ação

---

*[test/cucumber/support/groups.feature, line 25 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYA9Z5QNEf-RwzhTqf8:-KYA9Z5QNEf-RwzhTqf9:bd1a7v0) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/cucumber/support/groups.feature#L25)):*
> ```Cucumber
>   And O pesquisador "PauloSoares" com o cpf "51111111113" pertence ao grupo "BiologiaComputacional"
>   And Eu vejo os pesquisadores "VeronicaTeichrieb", "FernandoCastor" e "PauloSoares"
>   When Eu seleciono a opcao de criar para o grupo "CInUFPE" com os pesquisadores  "VeronicaTeichrieb", "FernandoCastor" e "PauloSoares"
> ```

para o grupo -> o grupo

---

*[test/functional/pages/CreateGroupPage.groovy, line 11 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYAA3oIH6a1z1MuJVVE:-KYAA3oIH6a1z1MuJVVF:blk4ugs) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/functional/pages/CreateGroupPage.groovy#L11)):*
> ```Groovy
> 
>     static at = {
>         title ==~ /Create ResearchGroup/
> ```

internacionalizar

---

*[test/functional/pages/CreateGroupPage.groovy, line 20 at r2](https://reviewable.io:443/reviews/pauloborba/rpa/15#-KYAA9qTQcdhs6IXZ56i:-KYAA9qTQcdhs6IXZ56j:b-440a8m) ([raw file](https://github.com/pauloborba/rpa/blob/6095a0a67a223b61d7c74dc2a61166878f3b7d43/test/functional/pages/CreateGroupPage.groovy#L20)):*
> ```Groovy
>     }
> 
>     boolean ViewResearcher(pesq, num) {
> ```

métodos devem começar com minúsculas. deveria buscar o pesquisador no HTML, não na base

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pauloborba/rpa/15)*

File: test/cucumber/steps/GroupsSteps.groovy [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    assert page.createNewGroup(arg1, pesqs)
+}
+
+Then(~/^Eu vejo que o grupo "([^"]*)" foi criado com os pesquisadores "([^"]*)", "([^"]*)" e "([^"]*)"$/) { String arg1, String arg2, String arg3, String arg4 ->
+    to GroupListPage
+    at GroupListPage
+    assert ResearchGroup.findByName(arg1).findResearcher(Researcher.findByName(arg2))
+    assert ResearchGroup.findByName(arg1).findResearcher(Researcher.findByName(arg3))
+    assert ResearchGroup.findByName(arg1).findResearcher(Researcher.findByName(arg4))
+}
```
Comment (macm3) [inline test/cucumber/steps/GroupsSteps.groovy:132]: Faltou colocar a linha vazia abaixo.

File: test/cucumber/steps/GroupsSteps.groovy [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+import rpa.ResearchGroup
+import rpa.ResearchGroupController
+import rpa.Researcher
+import rpa.ResearcherController
+
+this.metaClass.mixin(cucumber.api.groovy.Hooks)
+this.metaClass.mixin(cucumber.api.groovy.EN)
+
+def controladorGrupo = new ResearchGroupController()
+def controladorPesquisador = new ResearcherController()
```
Comment (macm3) [inline test/cucumber/steps/GroupsSteps.groovy:15]: Criou um controlador de pesquisador que não está sendo usado para nada no código.

**Comment by macm3** [comment]
Faltou o arquivo env.groovy

**Comment by macm3** [comment]
Poderia ter usado Extract Class nas Pages para criar uma classe de página já com a internacionalização com as pages como extensão dela. Para evitar repetir o código do InternationalizationHelper.

**Pull request merged**